### DISCUSSION TITLE
Animations for player to player interactions and farming

### DISCUSSION THREAD

**Discussion by MarvinVomberg** [EveronLife/EveronLife#143] [Ideas]
We are currently planning the player-to-player interactions and animations play a big role in that. We need people that dive into the animation system and find out how a few things are done and what our options will be for adding custom animations to do the following things:

- Hands over the head (surrender)
- Being tied up while standing and lying (police arrest with handcuffs)
- Getting Tazered
- Sitting on benches and on the floor
- Lying on the floor
- Picking/collecting items from bushes and trees
- Hitting rocks/ores with a pickaxe
- Cutting trees with an axe
- Shoveling with a shovel
- Repairing vehicles / a general "doing work on something in front of the player"
- Woundcare (self or on others)
- Lying "unconscious" on the ground

The following questions should be clarified:

- Which animations already exist in Reforger and can be used for the points listed above?
- How to create full body animations (e.g. hands in the air) with central player graph override?
- How do we add "fake" full body animations e.g. with an invisible item in hand to make use of the per item animation injection system?
- Can the "item in hand holding pose" adjust all bones of the body or just the hands / arms?
- Are there other ways to add animations?

If you are interested in collecting info how the animation system works and making proof of concepts for any of the interactions listed please reply below.

**Comment by Vergyy** [comment]
For "Lying "unconscious" on the ground" would that be an animation or just a pose?

**Reply by 12gabriel3** [reply]
I think it would need to be an animation otherwise the character would teleport from being up to being unconscious

**Reply by KatataFish17** [reply]
For lying unconscious or being tazered we could possibly use the method
`CharacterControllerComponent controller;
controller.Ragdoll();`
This will put you in a ragdoll-state without killing you. However there's no vanilla method to undo this. A possible solution could be spawning a new Character and give him all of his previous inventory.
I've tried it with `GenericEntity CustomRespawn(int playerId, string prefab, vector position, vector rotation = vector.Zero){}` (inside of SCR_RespawnSystemComponent.c).
But I haven't got it to work sadly.

**Reply by Vergyy** [reply]
You can go unconscious from multiple stances such as standing, crouching, kneeling, etc. If it had to be an animation then you would need multiple animations which seems a little much does it not?

**Reply by Arkensor** [reply]
I am not sure. I was hoping that AR comes with some of this for wounded state ... No idea how far the medical system is, someone needs to dig into that.

**Reply by 12gabriel3** [reply]
> You can go unconscious from multiple stances such as standing, crouching, kneeling, etc. If it had to be an animation then you would need multiple animations which seems a little much does it not?

That's what the devs did, one animation for when you are crouching and get hit from behind, one for when you are standing and get hit from the side. You get the gist. They have tables full of animations

**Comment by 12gabriel3** [comment]
# Requirements

I recommend reading about how animations work on the engine, what is an animation graph, a workspace and so on [here](https://community.bistudio.com/wiki/Arma_Reforger:Animation_Editor) if you want a deeper understanding.
To make animation graphs you need to understand each [node](https://community.bistudio.com/wiki/Arma_Reforger:Animation_Editor:_Nodes) of a graph as well.
You also need to know how to make animations in blender. To create character animations use [this rig](https://github.com/BohemiaInteractive/Arma-Reforger-Samples/blob/main/SampleMod_NewCharacter/Assets/Characters/SampleCharacter/Character_AnimationRig_Example.blend) provided by BI. You also need [Enfusion Blender tools](https://community.bistudio.com/wiki/Arma_Reforger:Enfusion_Blender_Tools) to export the animation. 

# How it works

You make animations is blender, import them to the engine and put them in an [anim set aka Anim Instance](https://community.bistudio.com/wiki/Arma_Reforger:Animation_Editor#Anim_Set). It is basically a table or tables full of `.anm` files.
Each `.anm` file can also trigger event such as sounds.
With an [animation graph](https://community.bistudio.com/wiki/Arma_Reforger:Animation_Editor#Animation_Graph) you select which animation from the anim set should be played based on conditions or randomness. You can also combine the animations together.
A [workspace](https://community.bistudio.com/wiki/Arma_Reforger:Animation_Editor#Workspace) holds all this stuff. A workspace can have multiple anim sets and multiple graphs.

# How to add animations

## Character animations attached to items

The main control graph selects all the animations from crouching, running, and so on, and provides [Attachments](https://community.bistudio.com/wiki/Arma_Reforger:Animation_Editor:_Nodes#Attachment) that let you put your animations in the main graph. I found 3 different attachments: `Gadget`, `WeaponAttach` and `Grenade`. Each attachment is in a different position of the graph, but you can do full body animations on any afaik. To attach the animation you made, add an `InventoryItemComponent` to your prefab, and add an `ItemAnimationAttributes`.  The only thing I understand are Animation Attachments. Take this image as an example:

![image](https://user-images.githubusercontent.com/16885097/178618402-f1e39204-3824-4cf5-bb4c-48a39ee00104.png)

It says: 
* Take the anim graph `Flashlight_soviet01.agr`, starting at the node called `MasterControl` and put it in place of the attachment called `Gadget` in the main control graph
* Use the animations contained in the anim set called `Flashlight_soviet01_player.asi`

It then plays based on the main graph conditions. In the case of gadgets they happen when you show the item. Take a look at the `Locomotion.agr` graph in `player_main.aw` workspace.

### Animated items

I did not do any animated item but looking at the grenades as an example it seems you need a `BaseItemAnimationController` and bind the animation from there instead of from `InventoryItemComponent`.

## Gestures

Gestures dont have an attachment node. The only gesture we have now is pointing. You can see how it works at `Gestures.agr` graph in `player_main.aw` workspace.

### Adding a new gesture

To add a gesture we need to modify `Gestures.agr`. The gestures have an ID attached to them. You should add a new state to `Gestures.agr` in the [State Machine](https://community.bistudio.com/wiki/Arma_Reforger:Animation_Editor:_Nodes#State_Machine) with `IsCommand(CMD_Gestures) && GetCommandI(CMD_Gestures) == N` as the state condition where `N` is the ID of the gesture you want to add. 
Then you need to configure which key the gesture should be bound to in `Configs/System/chimeraInputCommon.conf`. Add an action called something useful, lets say `HaltGesture` and bind it to a key, lets say H. The gestures are triggered by using the method `TryStartCharacterGesture` in `CharacterControllerComponent`. You can trigger the gesture with an action listener in `InputManager`. Our code would look something like this:

```cs
enum EL_ECharacterGestures
{
    HALT = 2  // this number is the same as you checked on the state machine
}

modded class SCR_PlayerController
{
     override void OnPostInit(IEntity owner)
     {
         super.OnPostInit(owner);
         InputManager inputManager = GetGame().GetInputManager();
         inputManager.AddActionListener("HaltGesture", EActionTrigger.DOWN, OnHaltGesture);
         inputManager.AddActionListener("HaltGesture", EActionTrigger.UP, OnHaltGesture);
     }
     
     void OnHaltGesture( float value = 0.0, EActionTrigger reason = 0 )
	{
		if ( reason == EActionTrigger.DOWN )
		{
			m_CharacterController.TryStartCharacterGesture(EL_ECharacterGestures.HALT);
		} 
		else if ( reason == EActionTrigger.UP )
		{
			m_CharacterController.StopCharacterGesture();
		}
	}	
```

# Creating an animation

After you made your animation(s), export it using Enfusion Blender Tools.
![image](https://user-images.githubusercontent.com/16885097/178614120-cecb155e-fd38-4fa0-853b-7b9af84cd002.png)

Each animation you made in blender will be listed here
![image](https://user-images.githubusercontent.com/16885097/178615748-6c0b43a8-7b5f-4f44-a615-ac534b7bac25.png)

Check `Show exports` on the ones you want to export and add an export file.
![image](https://user-images.githubusercontent.com/16885097/178615985-6b1ed1fb-a9b9-499c-92f5-9acf1bd34cfd.png)

The animation type (full body, only left arm, etc) is set on the animation itself when you export it.
On the leftmost drop down menu, select the right export settings for what you want to do. (so many they dont fit on my screen)
![image](https://user-images.githubusercontent.com/16885097/178614264-381fa31b-c874-4703-81c4-f51ebdd8d44e.png)
Absolute (ABS) animations make the selected body parts do exactly what you animated, while I've had no success with additive (ADD) animations, the name implies that it "blends" your animation with what the character is doing.

Import the `.txa` files in the Workbench to get the `.anm` files. With the `.anm` files you can make the Anim set

**Comment by 12gabriel3** [comment]
Check out [this branch](https://github.com/12gabriel3/EveronLife/tree/gestures).
Go to the debug world and hold T. It will first holster your weapon and then if you hold again you will surrender. You cant do anything when you surrender, not even look around. I did it on purpose.
What I did was 
* Add a state to the gestures state machine with code 2. I had to override a bunch of things otherwise the animation would give me an error when trying to save.
* Add an action for surrendering
* Add a context when you are surrendered where you cant do anything.
* Add a component to take care of playing the surrender animation, holstering your weapon and activating the "frozen" context.

**Reply by Arkensor** [reply]
If we need to override so many core files to add a gesture then that will be a major problem. All of those become incompatible with game updates or especially other mods. I think we should try and see how far we can push an invisible item in hands to inject a full-body animation. That might be more effort initially but we stay compatible with other mods and server owner are able to add their own animations as well then.

**Reply by KatataFish17** [reply]
First of: great work so far!
A small suggestion: The keybind for surrendering being T is a little inconvenient. Since T is also used for talking by standard keybinding, a better solution would be key, which has not been binded yet. I've found U to be a possible alternative.

**Reply by 12gabriel3** [reply]
I just picked a random key for this example

**Comment by 12gabriel3** [comment]
@Arkensor 
I changed the implementation so there are [no overrides](https://github.com/12gabriel3/EveronLife/tree/gestures). You spawn an invisible item in the hand of the character and inject the animation into an Attachment. The animation must be configured per item and the exit pose is too fast, but we can probably work around that later.

**Reply by Arkensor** [reply]
Excellent, I think we can wrap the process of what to do with items already in hand into a small reusable utility to do something with them and then take the item to hand to play the animation and reverse the whole thing. Configuration work per item is not a big deal, we do it once for a few and then that's all good.

Q: What are your ideas to deal with in-hand items? You previously mentioned something in regards to holstering? Is that a thing? Can we always holster things or return them to inventory so the player has empty hands if we need that? I was not sure if your hands count as additional storage and your inventory would be "full" so you can't put them anywhere. Would of course be nice if that was not an issue. Otherwise, I thought about removing the items from the player briefly, putting them underground, and returning them back after the animation is completed - or adding an additional storage to the player that is "infinite" but not accessible to other systems to temporary store the items into so they stay on the character.

---

Have you tried to manipulate a full-body stance by this just for fun? Like moving legs into some position or a very rough "sitting down"? I am not sure if item animations can control the height of the player above ground like crouching can, so that would be nice to know.

Do you have an example clip of exciting too fast?

**Reply by 12gabriel3** [reply]
This shows the animation exiting too fast:
![Animation](https://user-images.githubusercontent.com/16885097/179334538-50a4fa73-f759-4beb-8ed4-2a18960010c6.gif)

This is the sitting animation:
![AnimationSitting](https://user-images.githubusercontent.com/16885097/179334757-fb058bf9-9339-4ace-b383-39da8a1477aa.gif)

I think we should wait next update before deciding if we use items to inject animations. They are going to change how modders can interact with the animation system, according to [Dev Report #8](https://reforger.armaplatform.com/news/dev-report-8):

>We are also working on a new script API and methods to modify animation logic on characters, we want you to be able to create your own animation commands and communicate with the animation system.

**Reply by Arkensor** [reply]
We should have a look at how the game handles the transition back from item hold pose. Maybe we can do something to change how it's held to invert the animation basically and then only at the very end delete the item.

I am not sure if the *next* update already contains anything for animations. We can wait, it's only one week away. In the meantime, we can check what vanilla animations are available.

If in-hand items are just not cutting it for some of the things we want to do then can claim the main file override for our framework. It would mean that individual servers are unable to add anything though and they could not use mods that have advanced animations - but then again those mods would claim exclusivity as well and become incompatible with other mods :/

**Comment by 12gabriel3** [comment]
I've been thinking of a system to do player interactions. I've come up with a concept I would like to discuss.

I want to define a new concept called "Character state". The name is not really important. Example of character states are:
* Surrendered
* Restrained
* Unconscious
* Being carried/lead
* Carrying someone
* Poisoned

 Each character state could set:
1. The animation that should be played, if any
2. The input context. Someone unconscious shouldn't be able to move, for example
3. Screen effects
4. Audio effects
5. The position of a second character
6. Anything, really

The character has a component that manages states. This component must be general enough that it doesnt need to be modified if you add more states.

The state can be triggered any time by:
* Itself, ie when listening for user inputs
* By the manager, ie when a user action or another component summons the manager

The code could look something like this:

```cs

enum EL_CharacterStates
{
     UNCONSCIOUS,
     SURRENDERED,
     RESTRAINED,
     CARRIED,
     CARRYING
}
class EL_BaseCharacterState : ScriptAndConfig
{
    bool m_IsActive = false;
    
    void OnPostInit(IEntity owner); // Used for registering listeners to player inputs or anything else

    // Called when the state is activated, can have parameters, such as the controlled character, strength of the poison effect etc
    void Activate(void param1 = null, void param2 = null, void param3 = null, void param4 = null, void param5 = null, void param6 = null);
    
    EL_CharacterStates GetType();
    
    bool IsActive() { return m_IsActive }
}

class EL_CharacterStateComponent : ScriptComponent
{
     [Attribute()]
     ref array<ref EL_BaseCharacterState> m_AvailableStates;

    void OnPostInit(IEntity owner)
    {
        foreach(EL_BaseCharacterState state : m_AvailableStates)
        {
            state.OnPostInit(owner);
        }
    }
    
    void ActivateState(EL_CharacterStates type, void param1 = null, void param2 = null, void param3 = null, void param4 = null, void param5 = null, void param6 = null)
    {
        Find(type).Activate(void param1, void param2, void param3, void param4, void param5, void param6);
    }
    
    EL_BaseCharacterState Find(EL_CharacterStates type)
    {
         // finds the state
    }
    
    bool IsStateActive(EL_CharacterStates type)
    {
         return Find(type).IsActive();
    }
}
```

Then each state could do whatever, equip an item to trigger an animation, decrease character health every second...
An action could call ActivateState with the controlled character as the parameter, then when the state gets activated it knows what to do
Actions could have a condition of being shown if a specific state is active

What do you guys think?

**Reply by Arkensor** [reply]
Given the count of "states" we want (I would add sitting as well) it makes sense to me to make it re-usable and abstract the activation and deactivation away yes. Animations like these can be expressed as states, some of them would likely benefit from screen effects etc yes:
- Hands over the head (surrender)
- Being tied up while standing and lying (police arrest with handcuffs)
- Getting Tazered
- Sitting on benches and on the floor
- Lying on the floor
- Lying "unconscious" on the ground

And all the other animations like:
- Picking/collecting items from bushes and trees
- Hitting rocks/ores with a pickaxe
- Cutting trees with an axe
- Shoveling with a shovel
- Repairing vehicles / a general "doing work on something in front of the player"
- Woundcare (self or on others)
Would generally be context-based and temporary. Most of them have an item attached to do the interaction anyway (pickaxe, shovel, axe, bandages, repair kit). What we do with bare hand gathering to animate that I am not sure but I have hopes that we can maybe use vanilla animations there somehow and not need to do anything manually anyway.

---

I suppose we would net sync the state of a character via this manager so we can also support JIP. So we make use of the good old pattern of which state has been set and Replication.Time() of when that happened. Or does the animation system by itself already sync somehow? Are animations controlled/initiated on the client or is this all done on the authority?

---

Carrying other persons ... hmmm I think that might be a bit too much. I could imagine player A going into a strange animation pose above ground etc and then being linked AddChild to the other person that enters an animation to hold his arms to carry the other on the shoulder, but this looks like a lot of effort to me. And not sure how well this would work with the player controller and especially the player camera. We will see. I would respect the possibility in the design of a general "two people work together on an animation" but not go in too deep on that yet.

**Reply by 12gabriel3** [reply]
> Carrying other persons ... hmmm I think that might be a bit too much.

I dont mean carrying as in a fireman carry, I mean grabbing the person, taking over the controls and moving the restrained character into a car or to prison. With the user action we could get the restrained character and the character which is grabbing it. Then we somehow we make one character follow the other one, could be with add child or any other way.

**Reply by 12gabriel3** [reply]
> Sitting on benches and on the floor

I think we could make sitting on benches a dummy vehicle with a compartment manager and animations

**Reply by Arkensor** [reply]
> I dont mean carrying as in a fireman carry

Oh, I was thinking about putting an injured person over my shoulder to carry them somewhere. But yeah, simply force-moving a player around is something we need. Not sure how well addchild works with one player vs the other. That could be tested in a little PoC via peertool.

> I think we could make sitting on benches a dummy vehicle with a compartment manager and animations

Not a bad idea, yeah we could make vehicles that only have the absolute minimum required setup and do something there. At least for benches. To lay down anywhere on the floor won't be possible, but the question is if we *need* that. I think for social aspects it might be enough to just be able to sit.

**Comment by Arkensor** [comment]
Has anyone found out yet how to maybe play an animation only during a long F hold animation? I have not yet been able to see bandages or anything like that in action, I would hope that instead of bandaging while holding F we would swing the axe or dig with the shovel. Once the action is complete or aborted the action should stop.

**Reply by 12gabriel3** [reply]
We could use these methods in `ScriptUserAction` to add and then remove the item from the hand of the player. The animation can be a loop.
```cs
//! Method called when the action is interrupted/canceled.
//! \param pUserEntity The entity that was performing this action prior to interruption
event void OnActionCanceled(IEntity pOwnerEntity, IEntity pUserEntity) { };
//! Method called from scripted interaction handler when an action is started (progress bar appeared)
//! \param pUserEntity The entity that started performing this action
event void OnActionStart(IEntity pUserEntity) { };
```

**Reply by Arkensor** [reply]
I was more talking about how the game already does it. Or does AR not have any "use" animation for bandages or food or anything?

**Reply by Arkensor** [reply]
Here for example. I would like to know how that bandage action played the animation for the time. In the `SCR_BandageUserAction` I can see some code to change the model appearance to the "unpacked" bandage and I see code for the "consumeable" effect apply `SCR_ConsumableBandage`, but that only removes the bleeding then. I am not sure how this is connected to the actual animation.

I used this code on action:
```
SCR_CharacterDamageManagerComponent damageManager = SCR_CharacterDamageManagerComponent.Cast(pUserEntity.FindComponent(SCR_CharacterDamageManagerComponent));
damageManager.AddRandomBleeding();	
```

https://user-images.githubusercontent.com/8494013/179347720-5e843aab-65f5-4bb5-b9f2-c0af77193a14.mp4

**Reply by 12gabriel3** [reply]
There is an animation command called CMD_BandageSelf that triggers the animation in the main graph

**Reply by Arkensor** [reply]
Do you see where it is being triggered in script? I do not see it amongst other commands in `CharacterCommandHandlerComponent`?

**Reply by 12gabriel3** [reply]
I suspect it is done via engine

**Comment by Arkensor** [comment]
Different topic: While we know how to add new animations, how does one go about looking through the existing ones? Is somebody able to scan what is available and see what currently listed use cases could be associated with them - or what new ideas based on existing animations come to mind?

**Reply by 12gabriel3** [reply]
Got to the animation editor, open existing workspaces and open the anim set. Each green dot is a different animation
![image](https://user-images.githubusercontent.com/16885097/179181940-20486dcf-b932-465b-a914-752217d7d83a.png)

**Reply by Arkensor** [reply]
And how do I play one of them like the death animation for example?

**Reply by 12gabriel3** [reply]
You would need to send a command to the main animation graph, Im not sure how to do that

**Reply by 12gabriel3** [reply]
I know you can trigger gestures pretty easily, as well as trigger character death, but I think triggering only the death animation is done engine side, im not really sure

**Comment by Vergyy** [comment]
Now that we know the "unconscious" state is coming in the second milestone update are there still going to be home made animations for this or are you guys going to wait?
